plugins/samples/hmac_token_validation/plugin.cc

```diff
@@ -0,0 +1,189 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START serviceextensions_plugin_hmac_token_validation]
+#include <openssl/hmac.h>
+#include <string>
+#include <vector>
+#include <sstream>
+#include <iomanip>
+
+#include "absl/strings/escaping.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_split.h"
+#include "absl/time/clock.h"
+#include "proxy_wasm_intrinsics.h"
+
+class HmacValidationRootContext : public RootContext {
+public:
+  explicit HmacValidationRootContext(uint32_t id, std::string_view root_id)
+      : RootContext(id, root_id) {}
+
+  bool onConfigure(size_t) override {
+    LOG_DEBUG("HMAC validation plugin configured");
+    return true;
+  }
+  
+  // Replace with your desired secret key.
+  std::string secret_key_ = "your-secret-key";
+  // Token validity period in seconds (5 minutes).
+  const int64_t token_validity_seconds_ = 300;
+};
+
+// Validates incoming requests using HMAC authentication tokens.
+// The token must be provided in the Authorization header with format:
+// "HMAC timestamp:hmac_signature" where:
+// - timestamp: Unix timestamp when token was created
+// - hmac_signature: HMAC-MD5 signature of "METHOD:PATH:timestamp" using secret key
+class HmacValidationContext : public Context {
+public:
+  explicit HmacValidationContext(uint32_t id, RootContext* root)
+      : Context(id, root), root_(static_cast<HmacValidationRootContext*>(root)) {}
+
+  FilterHeadersStatus onRequestHeaders(uint32_t headers_size, bool end_of_stream) override {
+    LOG_DEBUG(absl::StrCat("Processing request with ", headers_size, " headers"));
+
+    // 1. Check if Authorization header exists
+    auto auth_header = getRequestHeader("authorization");
+    if (!auth_header || auth_header->view().empty()) {
+      sendLocalResponse(401, 
+                       "WWW-Authenticate: HMAC realm=\"api\"",
+                       "Missing Authorization header", 
+                       {});
+      return FilterHeadersStatus::StopIteration;
+    }
+
+    // 2. Verify Authorization header format (must start with "HMAC ")
+    std::string_view auth_value = auth_header->view();
+    const std::string prefix = "HMAC ";
+    
+    if (auth_value.size() < prefix.size() || 
+        !absl::EqualsIgnoreCase(auth_value.substr(0, prefix.size()), absl::string_view(prefix))) {
+      sendLocalResponse(400, 
+                       "", 
+                       "Invalid Authorization scheme. Use 'HMAC'", 
+                       {});
+      return FilterHeadersStatus::StopIteration;
+    }
+
+    // 3. Parse token parts (timestamp:hmac)
+    std::string token = std::string(auth_value.substr(prefix.size()));
+    std::vector<std::string> token_parts = absl::StrSplit(token, absl::MaxSplits(':', 1)); 
```

Comment:
destructuring this into separate local variables token_hmac and token_timestamp_str would make references later on more readable.